### ISSUE TITLE
Breaking change of `dataclasses.dataclass` comparison semantics in 3.13+

### ISSUE THREAD

**Issue by daskol** [python/cpython#128294]
# Bug report

#### Brief Description

Optimization done in #109870 changed semantic of `dataclass` comparison.

#### Description

The pseudo code below shows meaning of changes have been done in #109870. The semantic differs largely because of a shortcut in `__eq__` implementation of sequence-like containers (see [Objects/object.c][1]). The shortcut essentially does `self[i] is other[i]`. Consequently, method `__eq__` of `self[i]` is not evaluated for identical objects in 3.12 during `dataclasses.dataclass` comparison.

[1]: https://github.com/python/cpython/blob/v3.13.1/Objects/listobject.c#L3376-L3378

```python
def __eq__312(self, other):
    return astuple(self) == astuple(other)

def __eq__313(self, other):
    for lhs, rhs in zip(astuple(self), astuple(other)):
        if not lhs == rhs:
            return False
    else:
        return True
``` 

According [Python docs](https://docs.python.org/3.13/library/dataclasses.html) (citation below), v3.13 introduces breaking change since it does not consider fields as a tuples for dataclass comparison.

> eq: If true (the default), an [`__eq__()`](https://docs.python.org/3.13/reference/datamodel.html#object.__eq__) method will be generated. This method compares the class as if it were a tuple of its fields, in order. Both instances in the comparison must be of the identical type.

#### Test Case

```python
import numpy as np
from dataclasses import dataclass

@dataclass
class A:
    xs: np.ndarray

a = A(np.ones(3))
b = A(a.xs)

print(a == b)  # FAIL (3.13); OK (3.12).
# ValueError: Value The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

### CPython versions tested on:

3.13 (3.12)

### Operating systems tested on:

Linux

**Comment by sobolevn** [comment]
In 3.12 we used to do something like: `return (self.x, self.y) == (other.x, other.y)`, see https://github.com/python/cpython/blob/3a726be127b38ad568b100176c2de3112ae07b9d/Lib/dataclasses.py#L1085-L1094

Now, instead we do: `self.x == other.x and self.y == other.y`, which triggers explicit `__eq__`.

Problems:
- This is a breaking change
- Reverting this would be a breaking change for 3.13 users 😢

**Comment by daskol** [comment]
Indeed, we have the problem of two chairs here. From my perspective, the right decision would be reverting as soon as possible until major Linux distros do not adopt 3.13 widely.

In advance, I could give some context on a domain where the issue araises first. It is machine learning where data classes are used a lot for neural network representation and maintaining weight collections. Comparison is broken there because overridden `__eq__` has vector rather scalar semantic.

**Comment by picnixz** [comment]
> method __eq__ of self[i] is not evaluated for identical objects in 3.12 

Considering that we largely short-circuit equality in general, I think we should revert it. In addition, we are now out-of-sync with "This method compares the class **as if it were a tuple of its fields**, in order" (emphasis mine). While this can be a breaking change, we haven't updated the documentation, so maybe it hasn't been observed.

Most of the time, two objects that are identical (in terms of pointers) should also compare equal, independently of whether there is a custom `__eq__` or not, though this is my own opinion. If we're worried about constructing the tuple, instead of doing `x[i] == y[i]`, we could do `x[i] is y[i] or x[i] == y[i]` to emulate this behaviour but I wonder how performance would be affected.

cc @Yhg1s as the 3.13 RM.

**Comment by sobolevn** [comment]
I will send a fix ASAP.

**Comment by ericvsmith** [comment]
See also #120645. I don't think that we could back out the change at this point.

**Comment by daskol** [comment]
From perspective of [PEP-557][1], data classes should be compared as tuples since it explicitly states that _Data Classes can be thought of as “mutable namedtuples with defaults”_. I guess that making data classes as close as possible to named tuples was a primary design goal. For this reason, breaking change inroduced by #120645 was not the wisest decision.

[1]: https://peps.python.org/pep-0557/

**Comment by serhiy-storchaka** [comment]
On other hand, we rarely write `(self.x is other.x or self.x == other.x) and (self.y is other.y or self.y == other.y)` in hand-made classes. We usually simply write `self.x == other.x and self.y == other.y`. And `dataclasses` are intended to remove boilerplate for struct-like classes. `namedtuple` use identity check because they are tuples, and tuples use identity check because they are sequences and for consistency with the identity check in `__contains__`. But dataclasses usually are not sequences.

Perhaps in future we will add a per-field boolean option for using identity check.

**Comment by sobolevn** [comment]
I think that in this case (no revert) we need to update the docs to reflect the behaviour change. The OP has several examples of older wordings like:

```rst
   - *eq*: If true (the default), an :meth:`~object.__eq__` method will be
     generated.  This method compares the class as if it were a tuple
     of its fields, in order.
```

**Comment by daskol** [comment]
> But dataclasses usually are not sequences.

Indeed, dataclasses usually are not sequences but they are fancy sequences. At least, it was design goal according to [PEP-557][1]. Anyway, I'd like to hear @ericvsmith opinion on this matter since he initially drafted data classes.

Another consequence of this breaking change is compatibility with third party data class libraries (e.g. general-purpose `attrs` and `pydantic` or domain purpose `chex` or `flax.struct`).

- General purpose `attrs`.

  ```python
  from attrs import define
  
  @define
  class Point:
      x: float
      y: float
  
  a = Point(0.0, float('nan'))
  b = Point(0.0, a.y)
  assert a == b  # OK (3.12); FAIL (3.13).
  ```
- General purpose `pydantic`.

  ```python
  from pydantic.dataclasses import dataclass
  
  @dataclass
  class Point:
      x: float
      y: float
  
  a = Point(0.0, float('nan'))
  b = Point(0.0, a.y)
  assert a == b  # OK (3.12); FAIL (3.13).
  ```

From my perspective, new semantic of `__eq__` is viable change in general but it requires notice period. Third parties and a lot of users appeared in six years since PEP-557 rely on the existing semantics. It will take huge community effort to bring consistency.

[1]: https://peps.python.org/pep-0557/

**Comment by ericvsmith** [comment]
I think it's unfortunate that there wasn't a warning of this change. But at this point it has been released, and rolling it back is problematic. Maybe a better notice in the release notes would be enough?

I don't think the PEP comment _dataclasses of can be thought of as “mutable namedtuples with defaults”_ implies that they should be implemented with tuple semantics where possible. The statement is a very high level overview of the functionality.

@rhettinger : what are your thoughts?

**Comment by daskol** [comment]
@rhettinger Any comments?

**Comment by ericvsmith** [comment]
The changed behavior is being documented and tested in #138991.

**Comment by Hnasar** [comment]
> Another consequence of this breaking change is compatibility with third party data class libraries (e.g. general-purpose attrs and pydantic or domain purpose chex or flax.struct).

Actually, attrs _also_ made a similar breaking change in [24.1.0](https://www.attrs.org/en/stable/changelog.html#id13):
> The generated _eq_ methods have been sped up significantly by generating a chain of attribute comparisons instead of constructing and comparing tuples. This change arguably makes the behavior more correct, but changes it if an attribute compares equal by identity but not value, like float('nan'). https://github.com/python-attrs/attrs/issues/1310 

So now new attrs and new dataclasses agree, at least.
```py
from attrs import define

@define
class Point:
    x: float
    y: float

a = Point(0.0, float('nan'))
b = Point(0.0, a.y)
assert a == b  # OK (attrs 23.2.0); FAIL (attrs 24.1.0).
```

However, unlike [dataclasses.field](https://docs.python.org/3/library/dataclasses.html#dataclasses.field), [attrs.field](https://www.attrs.org/en/stable/api.html#attrs.field) allows you to pass per-field equality overrides which helps. But otherwise for dataclasses, what we have as a workaround is

```
    def __eq__(self, value: object) -> bool:
        if not isinstance(value, type(self)):
            return NotImplemented
        return astuple(self) == astuple(value)
```
